### PR TITLE
Adding parameter for minimum range for raycasting to clear obstacles

### PR DESCRIPTION
**Pull request by danieljeswin**
---

## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   |  #1809  |
| Primary OS tested on | Ubuntu 20.04 |
| Robotic platform tested on | Gazebo simulation of waffle (tb3_simulation_launch file in nav2_bringup package) |

---

## Description of contribution in a few bullet points
* Added minimum range parameter from which raycasting begins to clear obstacles (raytrace_min_range)
* Renamed the raytrace_range parameter to raytrace_max_range as it specifies the maximum range for raycasting.

## Description of documentation updates required from your changes
* Added new parameter and modified one, so need to add them to the default configs and documentation pages
* Raycasting to clear obstacles can now start from a minimum range, so will have to change documentation to reflect this

---

### CODE REVIEW

**Comment by SteveMacenski** [comment]
Oh wow that was quick, I'll try to get to this this week.

File: doc/parameters/param_list.md
Comment (SteveMacenski) [inline doc/parameters/param_list.md:124]: We should make this default to 0 so that current behavior persists on default parameters.

File: nav2_bringup/bringup/launch/tb3_simulation_launch.py
Comment (SteveMacenski) [inline nav2_bringup/bringup/launch/tb3_simulation_launch.py:132]: If this TODO isn't needed any more, we should change this in another PR. Lets not change unrelated things in this one.

File: nav2_costmap_2d/plugins/obstacle_layer.cpp
Comment (SteveMacenski) [inline nav2_costmap_2d/plugins/obstacle_layer.cpp:137]: Change default here to 0

File: nav2_costmap_2d/test/module_tests.cpp
Comment (SteveMacenski) [inline nav2_costmap_2d/test/module_tests.cpp:73]: These should be all caps to be known as globals

File: nav2_costmap_2d/test/module_tests.cpp
Comment (SteveMacenski) [inline nav2_costmap_2d/test/module_tests.cpp:70]: Adding some tests to costmap_2d is necessary to show this new feature is working as expected

File: nav2_costmap_2d/plugins/obstacle_layer.cpp
Comment (SteveMacenski) [inline nav2_costmap_2d/plugins/obstacle_layer.cpp:576]: This entire block from L576 to L595 can be deleted. If you change the arguments for `raytraceLine` to take in a maximum and a minimum range (converted into cell distances),  then [after this line](https://github.com/ros-planning/navigation2/blob/d3fc28af28df764a153c85434f8b30270e963622/nav2_costmap_2d/include/nav2_costmap_2d/costmap_2d.hpp#L414) you have a distance you can check if its lower than to exit without calling `bresenham2D`. 

Do the same with `updateRaytraceBounds`

File: nav2_costmap_2d/plugins/voxel_layer.cpp
Comment (SteveMacenski) [inline nav2_costmap_2d/plugins/voxel_layer.cpp:351]: I think you can do the same here as I suggested in the obstacle layer, but now with the voxel grid. Give the min range to `clearVoxelLineInMap` and `updateRaytraceBounds` and update once it itself computes the distances to return if too low.

**Review by SteveMacenski — CHANGES_REQUESTED** [review]
Good first stab, but I think you needed to go just 1 more layer lower to find a really clean place to add this without having to add any additional computations

File: nav2_costmap_2d/include/nav2_costmap_2d/costmap_2d.hpp
Comment (SteveMacenski) [inline nav2_costmap_2d/include/nav2_costmap_2d/costmap_2d.hpp:415]:
```suggestion
    if (dist < min_length) {
```

File: nav2_costmap_2d/include/nav2_costmap_2d/costmap_2d.hpp
Comment (SteveMacenski) [inline nav2_costmap_2d/include/nav2_costmap_2d/costmap_2d.hpp:423]: Move this line back please so all the offset calculations are together

File: nav2_costmap_2d/plugins/voxel_layer.cpp
Comment (SteveMacenski) [inline nav2_costmap_2d/plugins/voxel_layer.cpp:405]: Doesn't this need min range too to know whether it should touch the cells if its lower than the min distance?

File: nav2_voxel_grid/include/nav2_voxel_grid/voxel_grid.hpp
Comment (SteveMacenski) [inline nav2_voxel_grid/include/nav2_voxel_grid/voxel_grid.hpp:244]:
```suggestion
    if ((unsigned int)(dist) < min_length) {
```

File: nav2_costmap_2d/include/nav2_costmap_2d/costmap_2d.hpp
Comment (SteveMacenski) [inline nav2_costmap_2d/include/nav2_costmap_2d/costmap_2d.hpp:429]: I don't 100% understand the `- min_length` both here and in the 3D voxel layer, can you explain it? I would have thought you'd want to give it the min too, I think this offsets the raytracing by the min length (e.g. if max is 10 but in is 1, then I think it should only raytrace from 0-9 rather than 1-10). I don't see how this could know what the minimum range is to know to skip over it and start there to length

**Review by SteveMacenski — CHANGES_REQUESTED** [review]

File: nav2_costmap_2d/plugins/voxel_layer.cpp
Comment (SteveMacenski) [reply nav2_costmap_2d/plugins/voxel_layer.cpp:405]: both here and in the 2D layer

Comment (danieljeswin) [reply nav2_costmap_2d/plugins/voxel_layer.cpp:405]: From what I understand, updateRaytraceBounds function updates the maximum and minimum bounds of the maps. So in that case I think only the max range would be relevant for the map bounds since it is a line. Is my train of thought incorrect?

Comment (danieljeswin) [reply nav2_costmap_2d/plugins/voxel_layer.cpp:405]: From what I understand, updateRaytraceBounds function updates the maximum and minimum bounds of the maps. So in that case I think only the max range would be relevant for the map bounds since it is a line. Is my train of thought incorrect?

Comment (SteveMacenski) [reply nav2_costmap_2d/plugins/voxel_layer.cpp:405]: If in the lines above you are clearing a line, but the line is too short so you skip, I don't think that should affect the bounds of the map for update since that measurement was skipped.

Which actually reminds me that this logic is all good for clearing (e.g. if too short, don't clear), but you don't use the min range parameter in the marking side of things to skip marking if too close as well. It basically just needs something like this https://github.com/ros-planning/navigation2/blob/f86e44cedee5b6851e1aa09b20b09554d205b64a/nav2_costmap_2d/plugins/obstacle_layer.cpp#L390-L393 to check if too close (and `sq_dist` is already computed so no additional computations required)

Comment (danieljeswin) [reply nav2_costmap_2d/plugins/voxel_layer.cpp:405]: I mean that updating bounds with both max_range and min_range would be redundant since any updates with the max_range would override updates with the min_range

File: nav2_costmap_2d/include/nav2_costmap_2d/costmap_2d.hpp
Comment (danieljeswin) [reply nav2_costmap_2d/include/nav2_costmap_2d/costmap_2d.hpp:429]: I update x0, y0 and offset in the three lines above. That makes it start from min_range. Therefore the new length needs to be reduced by min_range

File: nav2_costmap_2d/plugins/voxel_layer.cpp
Comment (SteveMacenski) [reply nav2_costmap_2d/plugins/voxel_layer.cpp:405]: Ah, I see the miscommunication. I don't mean update the bounds with min range, I'm saying we should skip updating bounds entirely if the `min range > point range`, so that we should entirely ignore it since it fell short.

File: nav2_costmap_2d/include/nav2_costmap_2d/costmap_2d.hpp
Comment (SteveMacenski) [reply nav2_costmap_2d/include/nav2_costmap_2d/costmap_2d.hpp:429]: I see, thanks, can you add a comment to that effect in that calculation so that its clear to others later? Try to group those calculations together so that its not as spread out

**Comment by SteveMacenski** [comment]
There's still some open review comments that haven't been fixed yet :smile:

**Comment by danieljeswin** [comment]
I think they're all addressed now.

File: nav2_bringup/bringup/launch/tb3_simulation_launch.py
Comment (SteveMacenski) [inline nav2_bringup/bringup/launch/tb3_simulation_launch.py:132]: this is still changed, please revert.

File: nav2_costmap_2d/include/nav2_costmap_2d/costmap_2d.hpp
Comment (SteveMacenski) [inline nav2_costmap_2d/include/nav2_costmap_2d/costmap_2d.hpp:419]: Don't override variables with new quantities. Make a new variable and name it so this can be understood

File: nav2_costmap_2d/plugins/voxel_layer.cpp
Comment (SteveMacenski) [inline nav2_costmap_2d/plugins/voxel_layer.cpp:356]: remove space please

File: nav2_voxel_grid/include/nav2_voxel_grid/voxel_grid.hpp
Comment (SteveMacenski) [inline nav2_voxel_grid/include/nav2_voxel_grid/voxel_grid.hpp:250]: Same here on not re-using variables, it would be good to make new ones clear about what this is

File: nav2_voxel_grid/test/voxel_grid_tests.cpp
Comment (SteveMacenski) [inline nav2_voxel_grid/test/voxel_grid_tests.cpp:177]: Nice simple test for this, it would be good to have one for the obstacle layer's 2D as well

**Review by SteveMacenski — CHANGES_REQUESTED** [review]
Overall looks much better to me, very little additional overhead for the new feature! Though a little more cleanup is required

File: nav2_voxel_grid/test/voxel_grid_tests.cpp
Comment (SteveMacenski) [reply nav2_voxel_grid/test/voxel_grid_tests.cpp:177]: We want to make sure the project doesn't drop coverage % due to this addition

File: nav2_bringup/bringup/launch/tb3_simulation_launch.py
Comment (SteveMacenski) [reply nav2_bringup/bringup/launch/tb3_simulation_launch.py]: @danieljeswin please look at the diff on the PR. You keep not actually reverting this. Look what it was before you changed it and put that back in palce

File: nav2_costmap_2d/include/nav2_costmap_2d/costmap_2d.hpp
Comment (SteveMacenski) [inline nav2_costmap_2d/include/nav2_costmap_2d/costmap_2d.hpp:419]: how about `min_x0` and `min_y0`? New isn't much better than `temp`

File: nav2_costmap_2d/include/nav2_costmap_2d/costmap_2d.hpp
Comment (SteveMacenski) [inline nav2_costmap_2d/include/nav2_costmap_2d/costmap_2d.hpp:429]:
```suggestion
      // Subtract min_length from total length since initial point (x0, y0) has been adjusted by min Z
```

File: nav2_costmap_2d/include/nav2_costmap_2d/costmap_2d.hpp
Comment (SteveMacenski) [inline nav2_costmap_2d/include/nav2_costmap_2d/costmap_2d.hpp:440]:
```suggestion
    // Subtract min_length from total length since initial point (x0, y0) has been adjusted by min Z
```

File: nav2_system_tests/src/waypoint_follower/tester.py
Comment (SteveMacenski) [inline nav2_system_tests/src/waypoint_follower/tester.py:157]: Thanks for adding these

File: doc/parameters/param_list.md
Comment (SteveMacenski) [inline doc/parameters/param_list.md:123]: Please also update the names of these in the `nav2_bringup/config/*` files so that those have these new parameters reflected. 

We also need a PR in https://github.com/ros-planning/navigation.ros.org to reflect the parameter change in the costmap configuration pages and migration guide that this has changed.

Comment (SteveMacenski) [reply doc/parameters/param_list.md:123]: I found the pages for you for reference
- https://github.com/ros-planning/navigation.ros.org/blob/master/configuration/packages/costmap-plugins/obstacle.rst
- https://github.com/ros-planning/navigation.ros.org/blob/master/configuration/packages/costmap-plugins/voxel.rst
- https://github.com/ros-planning/navigation.ros.org/blob/master/migration/Foxy.rst

**Comment by SteveMacenski** [comment]
Just some last clean up and this can go in.

File: doc/parameters/param_list.md
Comment (SteveMacenski) [reply doc/parameters/param_list.md:123]: Add a new section to the migration page just explaining the parameter name change and that there's a minimum one too - link to this PR and the ROS answers ticket

Comment (danieljeswin) [reply doc/parameters/param_list.md:123]: Thanks for the info! I am not able to find any folder named config in the nav2_bringup package. I think I have updated all references to the old parameters (raytrace_range)

Comment (SteveMacenski) [reply doc/parameters/param_list.md:123]: Sorry, params, https://github.com/ros-planning/navigation2/tree/main/nav2_bringup/bringup/params

I use those interchangably. I should probably figure out which is the correct one...

Problems for another day :smile:

**Review by SteveMacenski — APPROVED** [review]
I think the big thing now is just documentation! Great work on this.

**Comment by SteveMacenski** [comment]
Still needs the param files in `nav2_bringup` changed but after that this can be merged. I looked over your docs and they look good too, just need a quick formatting change

**Comment by danieljeswin** [comment]
Made changes to the params file and corrected the formatting issue with the docs as well

**Review by SteveMacenski — APPROVED** [review]
LGTM, will merge when the docs is also merged

File: doc/parameters/param_list.md
Comment (SteveMacenski) [inline doc/parameters/param_list.md:123]: I thought we were going to do a min for both obstacle and ray trace? Then use the obstacle for the `marking` functions and the raytrace for the `clearing` functions - essentially decoupling them the same way the max ranges are

Comment (danieljeswin) [reply doc/parameters/param_list.md:123]: I'm not sure I understand what you mean. I created seperate min variables for the obstacle and raytracing. When I created the obstacle_min_range variable, I realized I could rename the obstacle_range to obstacle_max_range to maintain consistency

Comment (SteveMacenski) [reply doc/parameters/param_list.md:123]: But it looks like you also removed the other one. There are 2 max range parameters now and should have also 2 analog min parameters

Comment (danieljeswin) [reply doc/parameters/param_list.md:123]: No I have not removed the minimum range parameters

**Review by SteveMacenski — APPROVED** [review]
Perfect, no other issues on my end!

**Pull request merged**
